### PR TITLE
BitPacking functionality

### PR DESCRIPTION
**Pull request by mvaligursky**
A BitPacking namespace with functionality useful for packing bits into a number.

I've done some performance testing as well. The winner were free functions, followed by functions in an object (name space) - this was 17% slower, but still great. This followed by a class with static functions, which was massively slower, and so I've decided to not use that.

Screenshot 2023-02-10 at 11 20 09

### CODE REVIEW

File: src/core/math/bit-packing.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
+/**
+ * BitPacking API - functionality allowing to operate on values stored as bits in a number.
```
Comment (willeastcott) [inline src/core/math/bit-packing.js:2]:
```suggestion
* BitPacking API - functionality for operating on values stored as bits in a number.
```

File: src/core/math/bit-packing.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        return (storage & shifted) === shifted;
+    },
+
+    /**
+     * Tests if any specified bits are set.
+     *
+     * @param {number} storage - Number to test.
+     * @param {number} shift - Number of bits to shift the mask.
+     * @param {number} [mask] - Mask to limit the number of storage bits. Defaults to 1.
+     * @returns {number} Retuns true if any bits in the mask are set in the storage.
```
Comment (willeastcott) [inline src/core/math/bit-packing.js:57]:
```suggestion
 * @returns {boolean} Returns true if any bits in the mask are set in the storage.
```

File: src/core/math/bit-packing.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        return (storage >> shift) & mask;
+    },
+
+    /**
+     * Tests if all specified bits are set.
+     *
+     * @param {number} storage - Number to test.
+     * @param {number} shift - Number of bits to shift the mask.
+     * @param {number} [mask] - Mask to limit the number of storage bits. Defaults to 1.
+     * @returns {number} Retuns true if all bits in the mask are set in the storage.
```
Comment (willeastcott) [inline src/core/math/bit-packing.js:44]:
```suggestion
     * @returns {boolean} Returns true if all bits in the mask are set in the storage.
```

**Review by willeastcott — APPROVED** [review]
Approving with some minor suggestions. I notice this is being introduced without actually being used anywhere. What are the planned use cases? I assume all 4 functions are required for what you have planned?

**Comment by mvaligursky** [comment]
Yes, it's part of how I store render states - this will be a separate PR, just need to finish up. But this functionality looked a good one to be put into a separate PR.

**Comment by willeastcott** [comment]
I'm all for separate PRs. Nice job on the performance investigation too. We should always keep perf in mind.

**Pull request merged**